presentation/src/main/java/com/knocklock/presentation/lockscreen/LockScreenActivity.kt

```diff
@@ -0,0 +1,178 @@
+
+    private val manager by lazy {
+        this.applicationContext.packageManager
+    }
+
+    private val point by lazy { Point() }
+    private val composeViewModelStore by lazy { ViewModelStore() }
+    private val lifecycleOwner by lazy { ComposeLifecycleOwner() }
+    private var notificationListener: LockScreenNotificationListener? = null
+    private var mBound: Boolean = false
+    private val vm: LockScreenViewModel by viewModels()
+
+    private val connection = object : ServiceConnection {
+        override fun onServiceConnected(p0: ComponentName?, service: IBinder) {
+            val binder = service as LockScreenNotificationListener.LocalBinder
+            notificationListener = binder.getService()
+            mBound = true
+        }
+
+        override fun onServiceDisconnected(p0: ComponentName?) {
+            mBound = false
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        getWindowSize()
+        composeView = ComposeView(this).also { view ->
+            initViewTreeOwner(view)
+            view.setContent {
+                LockScreenHost(
+                    onFinish = {
+                        windowManager.removeViewImmediate(composeView)
+                        finishAndRemoveTask()
+                    },
+                    onRemoveNotifications = { keys ->
+                        if (mBound) notificationListener?.cancelNotifications(keys)
+                    },
+                    vm = vm,
+                    packageManager = manager,
+                )
+            }
+        }
+
+        windowManager.addView(composeView, getWindowManagerLayoutParams())
+    }
+
+    override fun onStart() {
+        super.onStart()
+        Intent(this, LockScreenNotificationListener::class.java).also { intent ->
+            bindService(intent, connection, Context.BIND_AUTO_CREATE)
+        }
+    }
+
+    override fun onStop() {
+        super.onStop()
+        unbindService(connection)
+        mBound = false
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        notificationListener = null
+        composeView = null
+    }
+    private fun initViewTreeOwner(composeView: ComposeView) {
+        lifecycleOwner.apply {
+            performRestore(null)
+            handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
+        }
+        ViewTreeLifecycleOwner.set(composeView, lifecycleOwner)
+        composeView.setViewTreeSavedStateRegistryOwner(lifecycleOwner)
+
+        ViewTreeViewModelStoreOwner.set(composeView) { composeViewModelStore }
+        val coroutineContext = AndroidUiDispatcher.CurrentThread
+        val runRecomposeScope = CoroutineScope(coroutineContext)
+        val recomposer = Recomposer(coroutineContext)
+        composeView.compositionContext = recomposer
+        runRecomposeScope.launch {
+            recomposer.runRecomposeAndApplyChanges()
+        }
```

Comment:
compose view 설정하는 코드군요.
처음보는 코드가 많아서 이것저것 찾아보다가 성빈랜드 블로그 글이 좋아서 가져왔어요.

읽다보니까 lifecycle이 onDestory될때 recomposer.close를 해줘야 한다는 글을 보았는데.
저희 코드에도 적용될 수 있을지.. 흠

 
https://sungbin.land/jetpack-compose-%EB%9F%B0%ED%83%80%EC%9E%84%EC%97%90%EC%84%9C-%EC%9D%BC%EC%96%B4%EB%82%98%EB%8A%94-%EB%A7%88%EB%B2%95-%EC%99%84%EC%A0%84%ED%9E%88-%ED%8C%8C%ED%95%B4%EC%B9%98%EA%B8%B0-recomposer-1455b2de3e20

Reply:
추가로 발견했는데 ON_START까지 설정해줘야 추후 LifeData 혹은 StateFlow가 제대로 동작할 것 같네요
https://stackoverflow.com/questions/72379865/jetpack-compose-recomposition-wont-trigger-when-using-custom-lifecycle-viewmod

Reply:
@jaeryo2357 기존에는 service에서  windowManager를 불러서 추가하는 형식이였어서, 위와 같은 코드의 예제를 많이 보았던 것 같아요 ( service에서는 lifecycle ON_CREATE만 handling해주었나 봅니다 ) 
지금은 activitiy로 옮겨왔기 때문에 첨부해주신 링크를 토대로 다시 작성해보아야할꺼같아요!

Reply:
@jaeryo2357 
```kotlin
@ExperimentalComposeUiApi
fun View.createLifecycleAwareWindowRecomposer(
    coroutineContext: CoroutineContext = EmptyCoroutineContext,
    lifecycle: Lifecycle? = null
): Recomposer {
    // Only access AndroidUiDispatcher.CurrentThread if we would use an element from it,
    // otherwise prevent lazy initialization.
    val baseContext = if (coroutineContext[ContinuationInterceptor] == null ||
        coroutineContext[MonotonicFrameClock] == null
    ) {
        AndroidUiDispatcher.CurrentThread + coroutineContext
    } else coroutineContext
    val pausableClock = baseContext[MonotonicFrameClock]?.let {
        PausableMonotonicFrameClock(it).apply { pause() }
    }

    var systemDurationScaleSettingConsumer: MotionDurationScaleImpl? = null
    val motionDurationScale = baseContext[MotionDurationScale] ?: MotionDurationScaleImpl().also {
        systemDurationScaleSettingConsumer = it
    }

    val contextWithClockAndMotionScale =
        baseContext + (pausableClock ?: EmptyCoroutineContext) + motionDurationScale
    val recomposer = Recomposer(contextWithClockAndMotionScale)
    val runRecomposeScope = CoroutineScope(contextWithClockAndMotionScale)
    val viewTreeLifecycle =
        checkNotNull(lifecycle ?: ViewTreeLifecycleOwner.get(this)?.lifecycle) {
            "ViewTreeLifecycleOwner not found from $this"
        }

    // Removing the view holding the ViewTreeRecomposer means we may never be reattached again.
    // Since this factory function is used to create a new recomposer for each invocation and
    // doesn't reuse a single instance like other factories might, shut it down whenever it
    // becomes detached. This can easily happen as part of setting a new content view.
    addOnAttachStateChangeListener(
        object : View.OnAttachStateChangeListener {
            override fun onViewAttachedToWindow(v: View) {}
            override fun onViewDetachedFromWindow(v: View) {
                removeOnAttachStateChangeListener(this)
                recomposer.cancel()
            }
        }
    )
    viewTreeLifecycle.addObserver(
        object : LifecycleEventObserver {
            override fun onStateChanged(
                lifecycleOwner: LifecycleOwner,
                event: Lifecycle.Event
            ) {
                val self = this
                when (event) {
                    Lifecycle.Event.ON_CREATE -> {
                        // Undispatched launch since we've configured this scope
                        // to be on the UI thread
                        runRecomposeScope.launch(start = CoroutineStart.UNDISPATCHED) {
                            var durationScaleJob: Job? = null
                            try {
                                durationScaleJob = systemDurationScaleSettingConsumer?.let {
                                    val durationScaleStateFlow = getAnimationScaleFlowFor(
                                        context.applicationContext
                                    )
                                    it.scaleFactor = durationScaleStateFlow.value
                                    launch {
                                        durationScaleStateFlow.collect { scaleFactor ->
                                            it.scaleFactor = scaleFactor
                                        }
                                    }
                                }
                                recomposer.runRecomposeAndApplyChanges()
                            } finally {
                                durationScaleJob?.cancel()
                                // If runRecomposeAndApplyChanges returns or this coroutine is
                                // cancelled it means we no longer care about this lifecycle.
                                // Clean up the dangling references tied to this observer.
                                lifecycleOwner.lifecycle.removeObserver(self)
                            }
                        }
                    }
                    Lifecycle.Event.ON_START -> pausableClock?.resume()
                    Lifecycle.Event.ON_STOP -> pausableClock?.pause()
                    Lifecycle.Event.ON_DESTROY -> {
                        recomposer.cancel()
                    }
                    Lifecycle.Event.ON_PAUSE -> {
                        // Nothing
                    }
                    Lifecycle.Event.ON_RESUME -> {
                        // Nothing
                    }
                    Lifecycle.Event.ON_ANY -> {
                        // Nothing
                    }
                }
            }
        }
    )
    return recomposer
}
```
내부의  클래스로 recomposer 객체를 만들면 민욱이가 올려준 부분을 다 적용할 수 있을꺼같아용~

Reply:
@jaeryo2357 [c6ba57a](https://github.com/Knock-Lock/Knock-Lock/pull/109/commits/c6ba57a31251867dd6101eb39bf829ef2d9624d7)  ViewTree설정코드를 다시 만들었습니당